src/commands/tail/mod.rs

```diff
@@ -1,15 +1,53 @@
+use std::net::{SocketAddr, TcpListener};
+
 use crate::settings::global_user::GlobalUser;
 use crate::settings::toml::Target;
 use crate::tail::Tail;
 
-pub fn start(target: &Target, user: &GlobalUser) -> Result<(), failure::Error> {
-    // Note that we use eprintln!() throughout this file; this is because we want any
+const DEFAULT_TUNNEL_PORT: u16 = 8080;
+const DEFAULT_METRICS_PORT: u16 = 8081;
+
+pub fn start(
+    target: &Target,
+    user: &GlobalUser,
+    tunnel_port: Option<u16>,
+    metrics_port: Option<u16>,
+) -> Result<(), failure::Error> {
+    let t_port = find_open_port(tunnel_port, DEFAULT_TUNNEL_PORT)?;
+    let m_port = find_open_port(metrics_port, DEFAULT_METRICS_PORT)?;
+
+    // Note that we use eprintln!() throughout this module; this is because we want any
     // helpful output to not be mixed with actual log JSON output, so we use this macro
     // to print messages to stderr instead of stdout (where log output is printed).
     eprintln!(
-        "Setting up log streaming from Worker script \"{}\" to Wrangler. This may take a few seconds...",
-        target.name
+        "Setting up log streaming from Worker script \"{}\". Using ports {} and {}. This may take a few seconds...",
```

Comment:
may want to change this copy to "up to a minute" or something? it's usually more than like 2-3 seconds

Reply:
i actually think i want to remove the copy and add the spinner #1161

Reply:
beautiful